### ISSUE TITLE
Unused Server-Only Dynamic Import Inside `typeof window === 'undefined'` Causes Error

### ISSUE THREAD

**Issue by karlhorky** [vercel/next.js#36514]
### Verify canary release

- [X] I verified that the issue exists in Next.js canary release

### Provide environment information

```bash
$ npx --no-install next info

    Operating System:
      Platform: linux
      Arch: x64
      Version: #112-Ubuntu SMP Thu Feb 3 13:50:55 UTC 2022
    Binaries:
      Node: 14.18.1
      npm: 6.14.15
      Yarn: 1.22.17
      pnpm: N/A
    Relevant packages:
      next: 12.1.6-canary.9
      react: 18.0.0
      react-dom: 18.0.0
```

### Describe the Bug

Hi there! First of all, thanks for your continued effort on Next.js! 🙌 Really amazing framework.

Importing a client-only function in a Next.js page causes bundling of the server-only code contained within the `typeof window === 'undefined'` check:

```js
// pages/index.js
import { usedClientExport } from "../util/shared";

export default function IndexPage() {
  usedClientExport();
  return <div>Hello World</div>;
}
```

```js
// util/shared.js
export async function unusedServerExport() {
  if (typeof window === "undefined") {
    await import("fs");
  }
}

export function usedClientExport() {
  console.log("yay");
}
```

Error:

<img width="690" alt="Screen Shot 2022-04-27 at 16 36 32" src="https://user-images.githubusercontent.com/1935696/165544897-19376853-ebaa-4b89-8627-390d8cad3c25.png">

```
./util/shared.js:3:10
Module not found: Can't resolve 'fs'
  1 | export async function unusedServerExport() {
  2 |   if (typeof window === "undefined") {
> 3 |     await import("fs");
    |          ^
  4 |   }
  5 | }
  6 | 
```

Sandbox: https://codesandbox.io/s/recursing-lalande-k3w986?file=/util/shared.js

cc @balazsorban44 @sokra 

### Expected Behavior

1. Any unused functions (in my example, `unusedServerExport`) are removed using tree shaking or dead code elimination and not included in the client bundle
2. Any code in `typeof window === 'undefined'` is not included in the client bundle
3. No errors appear for any code that will not be included in the bundle (in either of 1 or 2)

### To Reproduce

https://codesandbox.io/s/recursing-lalande-k3w986?file=/util/shared.js

‼️ This also occurs when using Babel instead of SWC:

https://codesandbox.io/s/unruffled-fire-jrkz7x?file=/pages/index.js (also here on StackBlitz: https://stackblitz.com/edit/nextjs-wmguir?file=pages%2Findex.js)

### Potentially Related Issues

- #30892
- #33791

**Comment by karlhorky** [comment]
### Workaround

A (very weird) workaround is to omit the `await` before the dynamic `import()`:

```js
export async function unusedServerExport() {
  if (typeof window === "undefined") {
    import("fs").then((m) => {
      console.log("imported", m);
    });
  }
}
```

Sandbox: https://codesandbox.io/s/empty-https-y9ws41?file=/util/shared.js:0-163

<img width="451" alt="Screen Shot 2022-04-27 at 16 49 28" src="https://user-images.githubusercontent.com/1935696/165546404-c0acad2c-7f90-4d86-a865-be1fa3cb20a2.png">

**Comment by timneutkens** [comment]
Seems the main difference between the Babel/SWC approach is that the Babel transform we used to transform `typeof window` actually evaluates the expression and inlines the result whereas with the SWC transform it doesn't evaluate it:

```js
// util/shared.js
export function unusedServerExport() {
  if (typeof window === "undefined") {
    import("fs");
  }
}

export function usedClientExport() {
  console.log("yay");
}
```

Result with SWC:

```js
// util/shared.js
export function unusedServerExport() {
    if ("object" === "undefined") {
        import("fs");
    }
}
export function usedClientExport() {
    console.log("yay");
}

```

Result with next/babel:

```js
// util/shared.js
export function unusedServerExport() {
    if (false) {
        import("fs");
    }
}
export function usedClientExport() {
    console.log("yay");
}
```

**Comment by karlhorky** [comment]
Ok, but two things:

1. In your code above, there is no `await` in front of the `import()`
2. The problem also appears when using Babel instead of SWC https://codesandbox.io/s/unruffled-fire-jrkz7x?file=/pages/index.js

**Comment by timneutkens** [comment]
Note that Next.js does not claim to even tree-shake these, we only tree shake get(Static/ServerSide)Props and getStaticPaths. For `typeof window` we only transform this to `'object'` or `'undefined'` and then Terser/swcMinify remove the code during minification. In this case webpack will still try to resolve/bundle all `import()` calls because well, they're in your code even when transformed. I'm not sure why a certain case you're referring to above would exclude it, maybe that's a bug even.

**Comment by karlhorky** [comment]
So am I understanding correctly that it's impossible to write a shared util file like this with Next.js?

By "shared util" I mean a file including:

1. Exported backend functions (depending internally on Node.js deps such as `fs`, with dynamic import)
2. Exported frontend functions (depending internally on frontend-only concepts like `localStorage`)

That would be too bad!

**Comment by karlhorky** [comment]
Or is it the current recommendation of the Next.js team to use the `isServer` and `config.node` webpack configuration to hack around issues like this? Seems a bit unusual to have to do this, especially when the code is not being used.

```js
module.exports = {
  webpack: (config, { isServer }) => {
    if (!isServer) {
      config.node = {
        fs: 'empty'
      };
    }
    return config;
  }
};
```

As seen in posts like [this](https://github.com/vercel/next.js/discussions/12124#discussioncomment-511175) or [this](https://javascript.plainenglish.io/understand-code-splitting-in-next-js-by-solving-a-common-error-db61a9b1b1a1)

**Comment by timneutkens** [comment]
> So am I understanding correctly that it's impossible to write a shared util file like this with Next.js?
> 
> By "shared util" I mean a file including:
> 
> 1. Exported backend functions (depending internally on Node.js deps such as `fs`, with dynamic import)
> 2. Exported frontend functions (depending internally on frontend-only concepts like `localStorage`)
> 3. Importing a frontend export in a shared component in `pages/` - outside of gSSP / gSP
> 
> That would be too bad! It's nice to group functions for both server and client in the same util file...

Correct, it's a compilation limitation. With Server Components there will be a file-level convention to mark something as server-only.

Related: https://github.com/vercel/next.js/issues/16153#issuecomment-673374488

**Comment by karlhorky** [comment]
Ok, maybe the `.server.js` / `.client.js` convention is something to adopt for non-React code then too, at least with Next.js apps, for now.

We'll also try out [the hacky-feeling workaround](https://github.com/vercel/next.js/issues/36514#issuecomment-1112074589) to see if this is any more ergonomic.

**Comment by stephenrobertsuk** [comment]
@karlhorky I also hit this problem when trying to create an isomorphic utility and am trying to work around it.  

So far I have managed by dynamically importing the module (which contains server-only code and imports) within a block which only executes in the server context (dependent on a ‘typeof window’ check). Also, adding this module to webpack externals if `!isServer`. This too feels a bit hacky in comparison to how this type of dead code may have been eliminated from the client bundle previously.

**Comment by pirtlj** [comment]
Also we need to ensure tree shaking so all the isomorphic server specific code does not get included in the client bundle

**Comment by mtsdalmolin** [comment]
Sorry for reviving this issue, but I don't understand what is the problema with my code. Is there anyway I could make this work?
I am trying to use `localStorage` inside a client component, but the compiler keeps yelling at me.

![image](https://github.com/vercel/next.js/assets/33668259/f9d5cc4f-eda9-4fd5-a280-b7a1c6292eed)

### components/test.client.js
```js
'use client';

if (localStorage === undefined) {
  console.log('im on server side');
}

export const Test = () => {
  return <h1>{JSON.stringify(localStorage)}</h1>;
};
```

### pages/index.js
```js
import { Test } from './components/test.client';

export default function IndexPage() {
  return <Test />;
}
```

StackBlitz reproduction:
https://stackblitz.com/edit/nextjs-bcwfdg?file=pages%2Findex.js,pages%2Fcomponents%2Ftest.client.js

**Comment by karlhorky** [comment]
Next.js 16.0.1 no longer has this issue 🎉

https://codesandbox.io/p/devbox/peaceful-blackwell-jftcg6?file=%2Futil%2Fshared.js%3A14%2C5&workspaceId=ws_GfAuHrswXyA1DoeSwsjjjz

<img width="1441" height="856" alt="Image" src="https://github.com/user-attachments/assets/2888781e-a617-4fbd-84d6-f6d4b2cb2d2f" />

Closing.
